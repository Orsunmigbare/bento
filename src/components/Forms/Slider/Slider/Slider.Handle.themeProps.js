import { themePropTypes } from "../../../../theming"

const themeProps = {
  handle: {
    name: "Slider_HANDLE",
    description: "Handle style",
    type: themePropTypes.style
  }
}
export default themeProps
