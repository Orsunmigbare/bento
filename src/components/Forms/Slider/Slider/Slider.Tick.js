/**
 * @class Slider
 */

import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import { applyVariants, color, font } from "../../../../theming"
import themeProps from "./Slider.Tick.themeProps"
import { componentCreateFactory } from "../../../../theming/utils/createVariantsFunctionFactory"

const SliderTick = ({ tick, count, format, className, ...restProps }) => (
  <div>
    <StartTick className={className} tick={tick} />
    <EndTick className={className} tick={tick} />
    {format(tick.value)}
  </div>
)

const StartTick = styled.button`
  position: absolute;
  left: ${p => p.tick.percent}%;
  margin-top: 14px;
  width: 1px;
  height: 5px;
  background-color: ${color.gray};
  ${applyVariants(themeProps.startTick)};
`
const EndTick = styled.button`
  position: absolute;
  left: ${p => p.tick.percent}%;
  margin-top: 22px;
  width: ${p => 100 / p.count};
  text-align: center;
  font-size: ${font.small};
  background-color: ${color.gray};
  ${applyVariants(themeProps.endTick)};
`

SliderTick.defaultProps = {
  className: ""
}

SliderTick.propTypes = {
  className: PropTypes.string,
  domain: PropTypes.array.isRequired,
  handle: PropTypes.shape({
    id: PropTypes.string.isRequired,
    value: PropTypes.number.isRequired,
    percent: PropTypes.number.isRequired
  }).isRequired,
  getHandleProps: PropTypes.func.isRequired
}
SliderTick.themeProps = themeProps
SliderTick.createVariants = componentCreateFactory(SliderTick)

/** @component */
export default SliderTick
