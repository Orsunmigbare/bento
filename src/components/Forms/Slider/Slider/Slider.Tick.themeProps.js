import { themePropTypes } from "../../../../theming"

const themeProps = {
  startTick: {
    name: "Slider_START_TICK",
    description: "Tick style",
    type: themePropTypes.style
  },
  endTick: {
    name: "Slider_END_TICK",
    description: "Tick style",
    type: themePropTypes.style
  }
}
export default themeProps
