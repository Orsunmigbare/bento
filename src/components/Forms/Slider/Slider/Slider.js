/**
 * @class Slider
 */

import React, { useState, useEffect } from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import {
  Slider as CompoundSlider,
  Rail,
  Handles,
  Tracks
} from "react-compound-slider"
import { applyVariants, spacing, color } from "../../../../theming"
import Handle from "./Slider.Handle"
import Track from "./Slider.Track"
import themeProps from "./Slider.themeProps"
import { componentCreateFactory } from "../../../../theming/utils/createVariantsFunctionFactory"

const Slider = ({
  variant,
  min = 0,
  max = 100,
  step = 10,
  onFocus,
  onKeyDown,
  onBlur,
  onUpdate,
  onChange,
  onSlideStart,
  defaultValue = 0,
  value,
  ...restProps
}) => {
  const [values, setValues] = useState([value || defaultValue])
  const [update, setUpdate] = useState([defaultValue])

  const handleUpdate = value => {
    setUpdate(value)
    onUpdate && onUpdate(update[0])
  }

  const handleChange = value => {
    setValues(values)
    onChange && onChange(value[0])
  }

  const handleStart = values => {
    setValues(values)
    onSlideStart && onSlideStart(values[0])
  }

  useEffect(() => {
    setValues([value])
  }, [value])

  return (
    <Container>
      <StyledSlider
        mode={1}
        step={step}
        domain={[min, max]}
        onUpdate={handleUpdate}
        onChange={handleChange}
        onSlideStart={handleStart}
        values={values}
        variant={variant}
        {...restProps}
      >
        <Rail>{({ getRailProps }) => <StyledRail {...getRailProps()} />}</Rail>
        <Handles>
          {({ handles, getHandleProps }) => (
            <div className="slider-handles">
              {handles.map((handle, index) => {
                const {
                  onKeyDown: handleKeyDown,
                  ...restHandleProps
                } = getHandleProps(handle.id)
                return (
                  <StyledHandle
                    key={handle.id}
                    handle={handle}
                    domain={[min, max]}
                    getHandleProps={getHandleProps}
                    variant={variant}
                    onFocus={onFocus}
                    onBlur={onBlur}
                    {...restHandleProps}
                    tabIndex={index}
                    onKeyDown={e => {
                      if (onKeyDown) {
                        onKeyDown(e.keyCode)
                      }
                      handleKeyDown(e)
                    }}
                  />
                )
              })}
            </div>
          )}
        </Handles>
        <Tracks right={false}>
          {({ tracks, getTrackProps }) => (
            <StyledTracks variant={variant}>
              {tracks.map(({ id, source, target }) => (
                <Track
                  key={id}
                  source={source}
                  target={target}
                  getTrackProps={getTrackProps}
                  variant={variant}
                />
              ))}
            </StyledTracks>
          )}
        </Tracks>
      </StyledSlider>
    </Container>
  )
}

const Container = styled.div`
  height: ${spacing.medium};
  padding-top: ${spacing.tiny};
  width: 100%;
  ${applyVariants(themeProps.container)};
`

const StyledSlider = styled(CompoundSlider)`
  position: relative;
  width: 100%;
  ${applyVariants(themeProps.slider)};
`

const StyledRail = styled.div`
  position: absolute;
  width: 100%;
  height: 14px;
  border-radius: 7px;
  cursor: pointer;
  background-color: ${color.gray};
  ${applyVariants(themeProps.rail)};
`

const StyledHandle = styled(Handle)`
  ${applyVariants(themeProps.handle)};
`

const StyledTracks = styled.div`
  ${applyVariants(themeProps.tracks)};
`

Slider.propTypes = {
  className: PropTypes.string,
  min: PropTypes.number,
  max: PropTypes.number,
  step: PropTypes.number,
  defaultValue: PropTypes.number,
  onChange: PropTypes.func,
  onKeyDown: PropTypes.func
}
Slider.themeProps = themeProps
Slider.createVariants = componentCreateFactory(Slider)

/** @component */
export default Slider
