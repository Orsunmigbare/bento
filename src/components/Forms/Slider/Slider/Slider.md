```jsx
<div>
  <Slider name="slider" onChange={() => console.log("Sliiiide")} />
</div>
```
