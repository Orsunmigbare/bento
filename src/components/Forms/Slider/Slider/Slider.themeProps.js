import {
  themePropTypes,
  createVariantsFunctionFactory
} from "../../../../theming"

const themeProps = {
  thumb: {
    name: "SLIDER_THUMB_STYLE",
    description: "Thumb style",
    type: themePropTypes.style
  },
  track: {
    name: "SLIDER_TRACK_STYLE",
    description: "Track style",
    type: themePropTypes.style
  },
  input: {
    name: "SLIDER_INPUT",
    description: "Input style",
    type: themePropTypes.style
  }
}

export default createVariantsFunctionFactory(themeProps)
