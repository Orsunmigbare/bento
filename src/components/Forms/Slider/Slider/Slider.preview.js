import React from "react"
import Slider from "./Slider"

const preview = {
  title: "Slider",
  category: "Components/Forms",
  component: Slider,
  render: props => (
    <div>
      <Slider
        name="slider"
        onChange={e => console.log("Sliiiide", e.target.value)}
        min={1}
        max={100}
      />
    </div>
  )
}

export default preview
