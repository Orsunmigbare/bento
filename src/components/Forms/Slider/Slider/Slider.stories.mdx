import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Slider from "./Slider"


<Meta title="Components|Forms/Slider" component={Slider}/>

# Slider
Slider is an input method where you set the value on something. Often used for tings like, volume, light or other similar things

```javascript
import { Slider } from "@staccx/bento"
```

## Standard Slider

<Preview>
  <Story name="Standard">
            <Slider
              name="Slider"
              onChange={e => console.log("Sliiiiide", e.target.value)}
              min={0}
              max={100}
              step={1}
            />
  </Story>
</Preview>

## Highter step
Slider has a step prop, that allowed you to controll how many steps it jumps when it moves. here it jumps 20 each time

<Preview>
  <Story name="High step">
            <Slider
              name="Step"
              onChange={e => console.log("Sliiiiide", e.target.value)}
              min={0}
              max={100}
              step={20}
            />
  </Story>
</Preview>

## Disabled

<Preview>
  <Story name="Disabled">
            <Slider
              name="Disabled"
              onChange={e => console.log("Sliiiiide", e.target.value)}
              min={0}
              max={100}
              disabled
            />
  </Story>
</Preview>



## Props for Slider

<Props of={Slider} />
