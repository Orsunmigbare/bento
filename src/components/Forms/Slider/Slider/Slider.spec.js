import React from "react"
import renderer from "react-test-renderer"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../../../theming/themes/baseTheme"
import Slider from "./Slider"

describe("Slider", () => {
  describe("Snapshots", () => {
    it("No props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Name", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider name="Slider" />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("onChange", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider name="Slider" onChange={e => console.log(e.target.value)} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("min", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider
              name="Slider"
              onChange={e => console.log(e.target.value)}
              min={100}
            />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("max", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider
              name="Slider"
              onChange={e => console.log(e.target.value)}
              min={0}
              max={100}
            />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Step", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider
              name="Slider"
              onChange={e => console.log(e.target.value)}
              min={0}
              max={100}
              step={20}
            />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("Step -", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider step={20} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("max -", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider max={100} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("min -", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider min={100} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("onChange -", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Slider onChange={e => console.log(e.target.value)} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
