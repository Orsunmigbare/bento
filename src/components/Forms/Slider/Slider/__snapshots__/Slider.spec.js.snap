// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Snapshots No props 1`] = `
<div
  className="sc-AxiKw emHJuz"
>
  <div
    className="sc-AxhCb jOJTnl"
    style={Object {}}
  >
    <div
      className="sc-AxhUy hyMSCF"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    />
    <div
      className="slider-handles"
    >
      <button
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        className="sc-AxjAm gTevRY sc-AxgMl cxJYYh"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
        type="button"
      />
    </div>
    <div
      className="sc-AxheI fcSpvz"
    >
      <div
        className="sc-AxirZ cESjww"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        target={
          Object {
            "id": "$$-0",
            "percent": 0,
            "value": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Slider Snapshots className 1`] = `
<div
  className="sc-AxiKw emHJuz"
>
  <div
    className="sc-AxhCb jOJTnl Slider"
    style={Object {}}
  >
    <div
      className="sc-AxhUy hyMSCF"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    />
    <div
      className="slider-handles"
    >
      <button
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={NaN}
        className="sc-AxjAm iGTrFZ sc-AxgMl cxJYYh"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
        type="button"
      />
    </div>
    <div
      className="sc-AxheI fcSpvz"
    >
      <div
        className="sc-AxirZ iWBtgN"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        target={
          Object {
            "id": "$$-0",
            "percent": NaN,
            "value": NaN,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Slider Snapshots max 1`] = `
<div
  className="sc-AxiKw emHJuz"
>
  <div
    className="sc-AxhCb jOJTnl"
    style={Object {}}
  >
    <div
      className="sc-AxhUy hyMSCF"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    />
    <div
      className="slider-handles"
    >
      <button
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        className="sc-AxjAm gTevRY sc-AxgMl cxJYYh"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
        type="button"
      />
    </div>
    <div
      className="sc-AxheI fcSpvz"
    >
      <div
        className="sc-AxirZ cESjww"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        target={
          Object {
            "id": "$$-0",
            "percent": 0,
            "value": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Slider Snapshots min 1`] = `
<div
  className="sc-AxiKw emHJuz"
>
  <div
    className="sc-AxhCb jOJTnl"
    style={Object {}}
  >
    <div
      className="sc-AxhUy hyMSCF"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    />
    <div
      className="slider-handles"
    >
      <button
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        className="sc-AxjAm gTevRY sc-AxgMl cxJYYh"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
        type="button"
      />
    </div>
    <div
      className="sc-AxheI fcSpvz"
    >
      <div
        className="sc-AxirZ cESjww"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        target={
          Object {
            "id": "$$-0",
            "percent": 0,
            "value": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Slider Snapshots step 1`] = `
<div
  className="sc-AxiKw emHJuz"
>
  <div
    className="sc-AxhCb jOJTnl"
    style={Object {}}
  >
    <div
      className="sc-AxhUy hyMSCF"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    />
    <div
      className="slider-handles"
    >
      <button
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        className="sc-AxjAm gTevRY sc-AxgMl cxJYYh"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
        type="button"
      />
    </div>
    <div
      className="sc-AxheI fcSpvz"
    >
      <div
        className="sc-AxirZ cESjww"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        target={
          Object {
            "id": "$$-0",
            "percent": 0,
            "value": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Slider Snapshots step 2`] = `
<div
  className="sc-AxiKw emHJuz"
>
  <div
    className="sc-AxhCb jOJTnl Slider"
    style={Object {}}
  >
    <div
      className="sc-AxhUy hyMSCF"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    />
    <div
      className="slider-handles"
    >
      <button
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={0}
        className="sc-AxjAm gTevRY sc-AxgMl cxJYYh"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
        type="button"
      />
    </div>
    <div
      className="sc-AxheI fcSpvz"
    >
      <div
        className="sc-AxirZ cESjww"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        target={
          Object {
            "id": "$$-0",
            "percent": 0,
            "value": 0,
          }
        }
      />
    </div>
  </div>
</div>
`;
